controllers/liveupdate_controller.go

```diff
@@ -0,0 +1,653 @@
+		}
+		return reconcile.Result{RequeueAfter: consts.DefaultWaitClusterNotWellFormed}, err
+	}
+	// If the new statefulSet was already created, check if the pods are ready
+	podList, err := r.podList(targetStatefulSet)
+	if err != nil {
+		r.reqLogger.Error(err, "Error getting pods of the new statefulSet '%s'", r.targetName)
+		return reconcile.Result{}, err
+	}
+
+	for _, pod := range podList.Items {
+		if !kube.IsPodReady(pod) {
+			r.reqLogger.Info(fmt.Sprintf("Pod '%s' not ready", pod.Name))
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, err
+		}
+		// Add a "version" label to pods
+		_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, &pod, func() error {
+			if pod.CreationTimestamp.IsZero() {
+				return errors.NewNotFound(corev1.Resource(""), pod.Name)
+			}
+			labels := pod.ObjectMeta.Labels
+			labels["version"] = strings.Split(consts.DefaultImageName, ":")[1]
+			pod.Labels = labels
+			return nil
+		})
+	}
+
+	// Move to the next stage
+	return r.updateStage(ispnv1.LiveUpgradeStart)
+}
+
+// prepare Prepares the new statefulSet, creating caches and connecting them to the source cluster
+func (r *LiveUpdateRequest) prepare() (ctrl.Result, error) {
+	// Obtain the podName from the source cluster to invoke requests
+	podListSource, err := PodList(r.infinispan, r.kubernetes, r.ctx)
+	if err != nil {
+		r.reqLogger.Error(err, "Failed to obtain pods from source cluster")
+		return reconcile.Result{}, err
+	}
+	podNameSource := podListSource.Items[0].Name
+
+	// Obtain the podName from the new stateFul to invoke requests
+	targetStatefulSet := &appsv1.StatefulSet{}
+	err = r.Get(r.ctx, types.NamespacedName{Namespace: r.infinispan.GetNamespace(), Name: r.targetName}, targetStatefulSet)
+	if err != nil && errors.IsNotFound(err) {
+		r.reqLogger.Error(err, "Failed to find the new statefulSet")
+		return reconcile.Result{}, err
+	}
+	podList, err := r.podList(targetStatefulSet)
+	if err != nil {
+		r.reqLogger.Error(err, "Failed to find the pods from the new statefulSet")
+		return reconcile.Result{}, err
+	}
+	podName := podList.Items[0].Name
+
+	// Obtain all cache names from the source cluster
+	names, err := r.cluster.CacheNames(podNameSource)
+	if err != nil {
+		r.reqLogger.Error(err, "failed to get cache names from the source cluster")
+		return reconcile.Result{}, err
+	}
+
+	r.reqLogger.Info(fmt.Sprintf("Cache names in the source cluster '%s'", names))
+	r.reqLogger.Info("Creating caches in the target cluster")
+
+	// Obtain the admin service in the source cluster to create the remote store config
+	sourceClusterService := &corev1.Service{}
+	res, err := kube.LookupResource(r.infinispan.GetAdminServiceName(), r.infinispan.GetNamespace(), sourceClusterService, r.Client, r.reqLogger, r.eventRec, r.ctx)
+
+	if res != nil {
+		r.reqLogger.Error(err, "failed to get service admin to connect to source cluster")
+		return reconcile.Result{}, err
+	}
+
+	for _, cacheName := range names {
```

Comment:
The logic from here on in the method concerns interactions with Infinispan, not k8s resources. Maybe a more appropriate place for this logic would be a method(s) in `remote_store.go`. This func could take `r.cluster`, `sourceClusterService` and `podName` as args and then perform the required actions there. That way we're separating the k8s and Infinispan logic. It also means that we don't have to return `reconcile.Result{}, err` etc for every stage of the upgrade.

The same applies for most of the cache logic in the `syncData` function.